### PR TITLE
seccomp: retry with mount hotplug

### PR DESCRIPTION
**Pull request by brauner**
When device creation via mknod fails, fallback to messing with mounts.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/seccomp.go [hunk 1]
```diff
@@ -441,6 +443,41 @@ func NewSeccompServer(d *Daemon, path string) (*SeccompServer, error) {
...
+		C.seccomp_notify_mknod_update_response(msg, 0)
+	}
+
+	return err
+}
+
+func doInjectDevice(dev types.Device, request_pid int, msg *C.struct_seccomp_notify_proxy_msg, d *Daemon) error {
+	c, err := findContainerForPid(int32(msg.monitor_pid), d)
+	if err == nil {
+		err = c.InsertSeccompUnixDevice("forkmknod.unix", dev, request_pid)
```
Comment (stgraber) [inline lxd/seccomp.go]: shouldn't we use the task's pid or something in there for when we eventually allow handling of multiple parallel requests for the same container?

File: lxd/container_lxc.go [hunk 2]
```diff
@@ -7749,7 +7792,7 @@ func (c *containerLXC) removeUnixDevices() error {
 	// Go through all the unix devices
 	for _, f := range dents {
 		// Skip non-Unix devices
-		if !strings.HasPrefix(f.Name(), "unix.") && !strings.HasPrefix(f.Name(), "infiniband.unix.") {
+		if !strings.HasPrefix(f.Name(), "forkmknod.unix.") && !strings.HasPrefix(f.Name(), "unix.") && !strings.HasPrefix(f.Name(), "infiniband.unix.") {
```
Comment (stgraber) [inline lxd/container_lxc.go]: Why are you making this change? If the container shutsdown and there's a leftover forkmknod.unix file, don't we want it deleted?

File: lxd/container.go [hunk 3]
```diff
@@ -729,6 +729,8 @@ type container interface {
 	CurrentIdmap() (*idmap.IdmapSet, error)
 	DiskIdmap() (*idmap.IdmapSet, error)
 	NextIdmap() (*idmap.IdmapSet, error)
+
+	InsertSeccompUnixDevice(prefix string, m types.Device, pid int) error
```
Comment (stgraber) [inline lxd/container.go]: Should go under the FIXME header above as this isn't a function we'll want to keep when we switch to the instance interface, so we'll need to change it to something else then.

File: lxd/container_lxc.go (hunk 2 above)
Comment (brauner) [reply lxd/container_lxc.go]: I don't understand. This is what this is doing.

File: lxd/seccomp.go (hunk 1 above)
Comment (brauner) [reply lxd/seccomp.go]: Not sure what exactly you're referring to here but request pid is the tasks pid.

Comment (stgraber) [reply lxd/seccomp.go]: In the file name

File: lxd/container_lxc.go (hunk 2 above)
Comment (stgraber) [reply lxd/container_lxc.go]: Oh, misread, nevermind

File: lxd/container.go (hunk 3 above)
Comment (stgraber) [reply lxd/container.go]: I meant move it up to be after DaemonState, no need to keep it at the end of the interface definition

File: lxd/seccomp.go (hunk 1 above)
Comment (brauner) [reply lxd/seccomp.go]: *shrug*, sure

File: lxd/seccomp.go [hunk 4]
```diff
@@ -441,6 +445,48 @@ func NewSeccompServer(d *Daemon, path string) (*SeccompServer, error) {
...
+	return nil, 0
+}
+
+func doInjectDevice(c container, dev types.Device, requestPID int) (error, int) {
+	err := c.InsertSeccompUnixDevice(fmt.Sprintf("forkmknod.unix.%d", requestPID), dev, requestPID)
+	if err != nil {
+		return err, int(-C.EPERM)
+	}
+
+	os.Remove(dev["hostpath"])
```
Comment (stgraber) [inline lxd/seccomp.go]: Any reason this can't be done as a defer in `InsertSeccompUnixDevice` instead?

**Comment by brauner** [comment]
jenkins: test this please

**Pull request merged**